src/ui/marker.js

```diff
@@ -268,16 +296,82 @@ class Marker {
             this._lngLat = smartWrap(this._lngLat, this._pos, this._map.transform);
         }
 
-        this._pos = this._map.project(this._lngLat)._add(this._offset);
+        const pos = this._pos = this._map.project(this._lngLat);
+
+        let anchor = this.options.anchor;
+        const offset = normalizeOffset(this.options.offset);
+
+        if (!anchor) {
+            const width = this._element.offsetWidth,
+                height = this._element.offsetHeight;
+
+            if (pos.y + offset.bottom.y < height) {
+                anchor = ['top'];
+            } else if (pos.y > this._map.transform.height - height) {
+                anchor = ['bottom'];
+            } else {
+                anchor = [];
+            }
+
+            if (pos.x < width / 2) {
+                anchor.push('left');
+            } else if (pos.x > this._map.transform.width - width / 2) {
+                anchor.push('right');
+            }
+
+            if (anchor.length === 0) {
+                anchor = 'bottom';
+            } else {
+                anchor = anchor.join('-');
+            }
+        }
```

Comment:
Similar to the discussion above re: offset, I'd be in favor of keeping the default behavior simpler, i.e., simply defaulting to `center` (since that's the current behavior)